### PR TITLE
enable _dnslink resolution

### LINKED ISSUE #5: Implement  _dnslink.example.com resolving

### PR DESCRIPTION
**Pull request by bonedaddy**
# Changes

* Enables the resolution of `_dnslink` records, and resolves [issue#5](https://github.com/ipfs/go-dnslink/issues/5). 
* Added a Travis CI build
* Enabled go modules
* Updated the readme to reference the correct DNSLink info website

# Notes

Right now this is using a fork of `jbenet/go-is-domain` since that repo is pretty stale, and I wasn't sure if/when this [PR](https://github.com/jbenet/go-is-domain/pull/8) would be merged. The fork also contains updated TLDs from IANA using version `2019031400`. 

Should the PR in `jbenet/go-is-domain` be merged, the dependencies for this PR will need to be updated.

### CODE REVIEW

**Comment by bonedaddy** [comment]
If desired I can include a fix for [issue#3](https://github.com/ipfs/go-dnslink/issues/3) in this PR as well.
For my own purposes I've written that functionality, see [here](https://github.com/RTradeLtd/go-dnslink/blob/bd9beebadb99913eff15d1a021e8252380b16457/dnslink.go#L107) if this would be something I should include here

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module github.com/ipfs/go-dnslink
+
+go 1.12
```
Comment (Stebalien) [inline go.mod]: Let's drop this to go 1.11 for now. We don't _yet_ require 1.12 (I think?).

**Review by Stebalien — COMMENTED** [review]
FYI, you should know that this library resolves /dns/..., not /ipns/... like IPFS does. Really, I believe we're supposed to be using `/dnslink` now by spec but we'll need to support both /ipns/... and /dnslink/... for the foreseeable future. If you want to use this library, you'll need to fix this.

**Comment by bonedaddy** [comment]
@Stebalien Ah okay thanks for the clarification, I was a little unsure as to the difference. For some additional clarification:

`/dns` (or more appropriately) `/dnslink` is used when resolving DNSLink  records, whereas `/dnsaddr`, `/dns4` and `/dns6` are used within libp2p to indicate that to get the IP of the peer to connect to, you must first resolve the DNS record to get the IP address?

Also the part of your comment ` If you want to use this library, you'll need to fix this.` I presume this means the library needs to be adjusted to handle `/dnslink` as well as `/dns` or rather, change to only support `/dnslink`

File: go.mod (hunk 1 above)
Comment (bonedaddy) [reply go.mod]: Done, I also changed travis to `1.11` to be consistent.

**Comment by bonedaddy** [comment]
As per your [comment](https://github.com/jbenet/go-is-domain/pull/8#discussion_r265819994) on the `go-is-domain` PR, the only place I can think of placing this would be [here](https://github.com/ipfs/go-dnslink/blob/4fc6e2c136b7f95f1f9150e6e70a0476f7d2c419/dnslink.go#L177).

The main issue I encountered when attempting to resolve `_dnslink` records was that `isDomain()` would trip on it, presumably because the regex doesn't allow use of `_`.

**Comment by Stebalien** [comment]
Ah, so, actually, that issue was a bit underspecified. It isn't about resolving `/dnslink/_dnslink.foo.com`. It's about resolving `/dnslink/foo.com` when the `dnslink` TXT record is attached to `_dnslink.foo.com`. That is, the resolution logic should be:

Given `/dnslink/foo.com`:

1. Lookup `_dnslink.foo.com`. If it has a TXT record with a value `dnslink=$PATH`, resolve to `$PATH`.
2. Lookup `foo.com` and follow the same procedure as above.

**Comment by bonedaddy** [comment]
Ah okay thanks for the clarification, I'll go ahead and make those changes :ok_hand:

**Comment by Stebalien** [comment]
Thanks! And thanks for picking this up!

**Comment by bonedaddy** [comment]
No problem happy to help! I'm about 99.98% sure that the last commits of mine added the correct functionality, let me know if there's anything i need to change. 

I also did a general find+replace to change any mentions of `/dns/` -> `/dnslink/`

**Comment by Stebalien** [comment]
I'm not seeing the `_dnslink` resolution part.

**Comment by bonedaddy** [comment]
Ah yes apologies. I believe the following `resolveOnce` change will satisfy the `_dnslink` resolution (unless im misunderstanding your comment), let me know if this works and i'll commit the change:

```Golang
// resolveOnce implements resolver.
func (r *Resolver) resolveOnce(domain string) (p string, err error) {
	r.setDefaults()
	// if given _dnslink.foo.com, trim _dnslink. before validating domain
	if !isd.IsDomain(strings.TrimPrefix(domain, "_dnslink.")) {
		return "", ErrInvalidDomain
	}
	txt, err := r.lookupTXT(domain)
	if err != nil {
		return "", err
	}
	err = ErrResolveFailed
	for _, t := range txt {
		p, err = ParseTXT(t)
		if err == nil {
			return p, nil
		}
	}

	return "", err
}
```

which results in the following output:

```shell
$ dnslink temporal.cloud
/ipfs/QmVqeSeC9DGVY2ui8VYE1dZ6EtyT63GqCDbkaniaiXgjGG
$ dnslink _dnslink.temporal.cloud
/ipfs/QmVqeSeC9DGVY2ui8VYE1dZ6EtyT63GqCDbkaniaiXgjGG
```

**Comment by Stebalien** [comment]
Sorry, that's not what I meant. The user should _never_ try to run `dnslink _dnslink.temporal.cloud`. The user will run `dnslink temporal.cloud` and the dnslink command should look for a dnslink TXT record on `_dnslink.temporal.cloud` and, if that doesn't work, on `temporal.cloud`.

That way, we can put the dnslink on the `_dnslink` subdomain and avoid polluting the root domain. The issue here is that the most specific DNS query we can make is "TXT". If everyone attaches their special TXT records to the root domain, the responses will get _massive_. That's why we support (and now encourage) putting the TXT record on the `_dnslink` subdomain (which isn't a valid hostname so it'll never conflict with any actual website subdomain).

**Comment by bonedaddy** [comment]
Ah I see thanks for the clarification. In that case, this should be the suitable solution

```Golang
// resolveOnce implements resolver.
func (r *Resolver) resolveOnce(domain string) (p string, err error) {
	r.setDefaults()
	if !isd.IsDomain(domain) {
		return "", ErrInvalidDomain
	}
	var txt []string
	// first attempt lookup on _dnslink.domain
	// if this fails, we'll then do a generalized
	// search against all TXT records for given domain
	// and if that fails, return an error
	txt, err = r.lookupTXT(fmt.Sprintf("_dnslink.%s", domain))
	if err != nil {
		txt, err = r.lookupTXT(domain)
		if err != nil {
			return "", err
		}
	}
	err = ErrResolveFailed
	for _, t := range txt {
		p, err = ParseTXT(t)
		if err == nil {
			return p, nil
		}
	}

	return "", err
}
```

**Comment by Stebalien** [comment]
That looks correct. Ideally, it would query both in parallel, preferring the result from `_dnslink.domain`, but we can handle that later.

**Comment by bonedaddy** [comment]
Sounds good! I've pushed the change

File: dnslink.go [hunk 2]
```diff
@@ -47,11 +47,12 @@ package dnslink
 
 import (
 	"errors"
+	"fmt"
 	"net"
 	"path"
 	"strings"
 
-	isd "github.com/jbenet/go-is-domain"
+	isd "github.com/RTradeLtd/go-is-domain"
```
Comment (Stebalien) [inline dnslink.go]: I've finished merging/fixing go-is-domain. Let's switch this back now.

Comment (bonedaddy) [reply dnslink.go]: Awesome, I pushed a new update that moved to juan's repository.

**Comment by bonedaddy** [comment]
Not quite sure why travis is failing to build after i switched the dependency to `jbenet/go-is-domain` :thinking:

**Comment by autonome** [comment]
@bonedaddy @Stebalien is this work still relevant? Any idea what the test failure is?

**Comment by lidel** [comment]
`/dns` → `/dnslink` and `_dnslink` subdomain support changes are still relevant to work @martinheidegger is doing, but it's been a while, and we are moving dnslink under https://github.com/dnslink-std, so we need to tackle those things in fresh PR.

<del>(lmk if you want to pick this up, or should we close it as stale and fill a new one)</del>

**Comment by martinheidegger** [comment]
:wave: Yes, and hi! This is still relevant. I have been working on a test suite for dnslink implementations and it currently [uses the `/dns/` prefix](https://github.com/dnslink-std/test/blob/58c1f7095769c86281e4dc7edb698a919c5f0953/integration-tests.js#L144) - the reason for this is because while most implementations simply ignore the prefix 😓, go-dnslink seems to be the only implementation (?0 that uses `/dns/` explicitly: https://github.com/ipfs/go-dnslink/blob/0626e50cfeaf0cd304eb0017de996e8cacb08540/dnslink.go#L220 - however it may be a good idea to support both as the documentation mentions `/dnslink/` everywhere. I opened an issue regarding this: https://github.com/dnslink-std/test/issues/9

**Pull request closed without merging**